pytorch_lightning/overrides/fairscale.py

```diff
@@ -0,0 +1,108 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Dict, Any, List, Tuple, Union
+
+from fairscale.nn.data_parallel.sharded_ddp import ShardedDataParallel
+from fairscale.optim import OSS
+from torch import nn
+
+from pytorch_lightning.utilities import rank_zero_warn
+
+
+class LightningOSS(OSS):
+
+    def state_dict(self) -> Dict[str, Any]:
+        """Return the last known global optimizer state, which consist of a list of the shards.
+
+        .. warning:
+            If the state has not been consolidated, this returns a shard's worth, not the global state.
+
+        .. warning:
+            Returning the global state is limited to the replica which was responsible for the consolidation.
+            The state may also not be up to date, depending on when `consolidate_state_dict` was last called.
+        """
+
+        if len(self._all_states) == 0:
+            rank_zero_warn("Optimizer state has not been consolidated. Returning the local state")
+            rank_zero_warn("Please call `consolidate_state_dict()` beforehand if you meant to save the global state")
+            state_dict = self.local_state_dict()
+            state_dict["local_state_dict"] = True
+            return state_dict
+
+        # Flatten the param_groups, save the partition which logs the rank <> shard correspondence
+        partition: List[Tuple[int, int]] = []
+        param_groups: List[Dict[Any, Any]] = []
+
+        start = 0
+        for i, s in enumerate(self._all_states):
+            param_groups.extend(s["param_groups"])
+            end = start + len(s["param_groups"])
+            partition.append((start, end))
+            start = end
+
+        return {
+            "state": [s["state"] for s in self._all_states],
+            "param_groups": param_groups,
+            "partition": partition,
+            "local_state_dict": False,
+        }
+
+
+class LightningShardedDataParallel(ShardedDataParallel):
+    def __init__(
+            self,
+            base_model: nn.Module,
+            sharded_optimizer: Union[OSS, List[OSS]],
```

Comment:
should the OSS wrapping be automatic? or should users explicitly use OSS in `configure_optimizers` of their lightning module?

Reply:
I think atm it supports both. We just skip wrapping the optimizer in the sharded_plugin if the user has already wrapped it!